File: experiment-5/runs/language=typescript_model=claude-opus-4-8_tooling=beads/rep3/src/queries/competitions.ts
/*
 * ============================================================================
 * Context
 * ----------------------------------------------------------------------------
 * Module:  src/queries/competitions.ts
 * Purpose: Competition-level queries: compute a league table (standings) for a
 *          given competition + season directly from match results, determine
 *          the champion, and (heuristically) the relegated teams.
 * Inputs:  A loaded `Dataset`, competition name, and season year.
 * Outputs: Sorted `StandingRow[]` and derived champion / relegation info.
 * Notes:   Standings use 3-1-0 points, tie-broken by points, goal difference,
 *          goals for, then name. Relegation heuristic = bottom 4 of a 20-team
 *          Série A table (only reported when the table has >= 16 teams).
 * ============================================================================
 */

import type { Dataset } from "../data/loader.js";
import type { StandingRow } from "../data/types.js";
import { findMatches } from "./matches.js";

/** Compute the full standings table for a competition + season. */
export function standings(
  ds: Dataset,
  competition: string,
  season: number,
): StandingRow[] {
  const matches = findMatches(ds, { competition, season });

  interface Acc {
    team: string;
    played: number;
    wins: number;
    draws: number;
    losses: number;
    goalsFor: number;
    goalsAgainst: number;
  }
  const table = new Map<string, Acc>();

  const ensure = (key: string, display: string): Acc => {
    let a = table.get(key);
    if (!a) {
      a = {
        team: display,
        played: 0,
        wins: 0,
        draws: 0,
        losses: 0,
        goalsFor: 0,
        goalsAgainst: 0,
      };
      table.set(key, a);
    }
    return a;
  };

  for (const m of matches) {
    if (m.homeGoals == null || m.awayGoals == null) continue;
    const home = ensure(m.homeKey, m.homeTeam);
    const away = ensure(m.awayKey, m.awayTeam);
    home.played++;
    away.played++;
    home.goalsFor += m.homeGoals;
    home.goalsAgainst += m.awayGoals;
    away.goalsFor += m.awayGoals;
    away.goalsAgainst += m.homeGoals;
    if (m.homeGoals > m.awayGoals) {
      home.wins++;
      away.losses++;
    } else if (m.homeGoals < m.awayGoals) {
      away.wins++;
      home.losses++;
    } else {
      home.draws++;
      away.draws++;
    }
  }

  const rows: StandingRow[] = [...table.values()].map((a) => ({
    position: 0,
    team: a.team,
    played: a.played,
    wins: a.wins,
    draws: a.draws,
    losses: a.losses,
    goalsFor: a.goalsFor,
    goalsAgainst: a.goalsAgainst,
    goalDifference: a.goalsFor - a.goalsAgainst,
    points: a.wins * 3 + a.draws,
  }));

  rows.sort(
    (x, y) =>
      y.points - x.points ||
      y.goalDifference - x.goalDifference ||
      y.goalsFor - x.goalsFor ||
      x.team.localeCompare(y.team),
  );
  rows.forEach((r, i) => (r.position = i + 1));
  return rows;
}

export interface CompetitionSummary {
  competition: string;
  season: number;
  champion: string | null;
  table: StandingRow[];
  relegated: string[];
}

/** Summarize a competition season: champion, full table, relegation guess. */
export function competitionSummary(
  ds: Dataset,
  competition: string,
  season: number,
): CompetitionSummary {
  const table = standings(ds, competition, season);
  const champion = table.length > 0 ? table[0].team : null;
  // Relegation heuristic for a points-based league with a full table.
  const relegated =
    table.length >= 16 ? table.slice(-4).map((r) => r.team) : [];
  return { competition, season, champion, table, relegated };
}

/** List the distinct seasons available for a competition. */
export function availableSeasons(ds: Dataset, competition: string): number[] {
  const seasons = new Set<number>();
  for (const m of findMatches(ds, { competition })) {
    if (m.season != null) seasons.add(m.season);
  }
  return [...seasons].sort((a, b) => a - b);
}

File: experiment-14/runs/language=elixir_model=opus-4.8-fast_prompt=TDD/rep1/lib/brazilian_soccer/queries/competitions.ex
defmodule BrazilianSoccer.Queries.Competitions do
  @moduledoc """
  Competition-level queries. League standings are computed from match results
  using the standard 3-points-for-a-win system, ranked by points, then goal
  difference, then goals scored, then name.
  """

  alias BrazilianSoccer.{Dataset, Match}
  alias BrazilianSoccer.Queries.{Matches, Source}

  @doc """
  Compute the standings table for a competition and season. Each entry includes
  position, team, played, wins/draws/losses, goals for/against, goal difference
  and points.
  """
  @spec standings(Dataset.t(), binary(), integer()) :: [map()]
  def standings(%Dataset{} = ds, competition, season) do
    matches =
      ds
      |> Matches.find(competition: competition, season: season)
      |> Source.primary_per_season()
      |> Enum.filter(&(Match.winner(&1) != nil))

    matches
    |> Enum.reduce(%{}, &accumulate/2)
    |> Map.values()
    |> Enum.map(&finalize/1)
    |> Enum.sort(&rank/2)
    |> Enum.with_index(1)
    |> Enum.map(fn {row, pos} -> Map.put(row, :position, pos) end)
  end

  @doc "Return the team at the top of the standings, or nil when empty."
  @spec champion(Dataset.t(), binary(), integer()) :: map() | nil
  def champion(%Dataset{} = ds, competition, season) do
    ds |> standings(competition, season) |> List.first()
  end

  @doc "Sorted list of seasons present for a competition."
  @spec seasons(Dataset.t(), binary()) :: [integer()]
  def seasons(%Dataset{matches: matches}, competition) do
    matches
    |> Enum.filter(&(&1.competition == competition and &1.season != nil))
    |> Enum.map(& &1.season)
    |> Enum.uniq()
    |> Enum.sort()
  end

  @doc "Sorted list of distinct competitions in the dataset."
  @spec competitions(Dataset.t()) :: [binary()]
  def competitions(%Dataset{matches: matches}) do
    matches
    |> Enum.map(& &1.competition)
    |> Enum.reject(&is_nil/1)
    |> Enum.uniq()
    |> Enum.sort()
  end

  defp accumulate(%Match{} = m, acc) do
    acc
    |> upsert(m.home_team, m.home_key, m.home_goals, m.away_goals)
    |> upsert(m.away_team, m.away_key, m.away_goals, m.home_goals)
  end

  defp upsert(acc, team, key, gf, ga) do
    row = Map.get(acc, key, blank(team))

    result =
      cond do
        gf > ga -> :wins
        ga > gf -> :losses
        true -> :draws
      end

    row =
      row
      |> Map.update!(:played, &(&1 + 1))
      |> Map.update!(:goals_for, &(&1 + gf))
      |> Map.update!(:goals_against, &(&1 + ga))
      |> Map.update!(result, &(&1 + 1))

    Map.put(acc, key, row)
  end

  defp blank(team) do
    %{team: team, played: 0, wins: 0, draws: 0, losses: 0, goals_for: 0, goals_against: 0}
  end

  defp finalize(row) do
    row
    |> Map.put(:goal_difference, row.goals_for - row.goals_against)
    |> Map.put(:points, row.wins * 3 + row.draws)
  end

  # Brazilian league tie-break order: points, then wins, then goal difference,
  # then goals for; finally team name ascending for stable display.
  defp rank(a, b) do
    {a.points, a.wins, a.goal_difference, a.goals_for, sortable(b.team)} >=
      {b.points, b.wins, b.goal_difference, b.goals_for, sortable(a.team)}
  end

  # Names sort ascending as a tie-breaker; invert by comparing b vs a above.
  defp sortable(name), do: String.downcase(name)
end
